lib/pbench/common/results.py

```diff
@@ -0,0 +1,119 @@
+import urllib.parse
+from logging import Logger
+from pathlib import Path
+
+import requests
+
+from pbench import PbenchConfig
+
+
+class CopyResultTb:
+    """Interfaces for copying result tar balls remotely using the server's HTTP
+    PUT method for uploads.
+    """
+
+    class FileUploadError(Exception):
+        """Raised when the uploading of file to server has failed"""
+
+        pass
+
+    def __init__(
+        self,
+        controller: str,
+        tarball: str,
+        tarball_len: int,
+        tarball_md5: str,
+        config: PbenchConfig,
+        logger: Logger,
+    ):
+        """Contructor for object representing tar ball to be copied remotely.
+
+        Raises
+            ValueError          if the given controller is not a valid hostname
+            FileNotFoundError   if the given tar ball does not exist
+        """
+        self.controller = controller
+        self.tarball = Path(tarball)
+        if not self.tarball.exists():
+            raise FileNotFoundError(f"Tar ball '{self.tarball}' does not exist")
+        self.tarball_len = tarball_len
+        self.tarball_md5 = tarball_md5
+        self.config = config
+        server_rest_url = config.get("results", "server_rest_url")
```

Comment:
Here's the problem:  the code here in `CopyRestultTb()` is going to be used by `pbench-results-put`/`pbench-results-move` and by `pbench-server-put-shim-001`, the former being Agent-side commands while the latter is a server, which means that the two uses are going to have very different configuration files.  Now, we can finesse this situation by having this function accept a `PbenchConfig` object instead of one of the subclasses, but this doesn't get around the fact that the _contents_ of the object will have to contain a common section (here, `results`) with a common field in it.  And, while this is an adequate constraint, it's an annoying one, especially since, per @npalaska, another field in the server configuration file _already_ has this info in it.

So, I propose that, instead of passing an entire config object into this function (from which the only field we use here is the server's REST URL), lets just pass in the URL, and let the caller of this function figure out how to get it.  That is, the caller apparently already has the config object (which they are passing in here), and the caller can know where in the object to look for the URL (i.e., the server can look in the server place while that agent can look in the agent place), and that saves this code from having to know where to look and saves us from having to make the server config look a little like the agent config.

Does that sound good to everybody?

Note that there is one catch:  `copy_result_tb()` is peeking in the config object to see if we're running in a unit test.  I'm hoping that it will stop doing that, in which case this dependency will go away....